core/cas-server-core-authentication-api/src/main/java/org/apereo/cas/authentication/AuthenticationCredentialsThreadLocalBinder.java

```diff
@@ -24,12 +27,31 @@
 
     private static final ThreadLocal<Authentication> CURRENT_AUTHENTICATION = new ThreadLocal<>();
 
+    private static final ThreadLocal<Principal> IN_PROGRESS_PRINCIPAL = new ThreadLocal<>();
+
     private static final ThreadLocal<Authentication> IN_PROGRESS_AUTHENTICATION = new ThreadLocal<>();
 
     private static final ThreadLocal<AuthenticationBuilder> CURRENT_AUTHENTICATION_BUILDER = new ThreadLocal<>();
 
     private static final ThreadLocal<String[]> CURRENT_CREDENTIAL_IDS = new ThreadLocal<>();
 
+    /**
+     * Bind Principal to ThreadLocal for principal that has been authenticated but for which attributes have not yet been retrieved.
+     * This can make attributes from the authentication event (e.g. X509, LDAP, WSFED) available to scripted attribute repositories.
+     * If a principal is already bound, combine the attributes with new attributes overriding existing.
+     * CAS won't use this as the principal, this principal is just an attribute bucket that will start with authentication
+     * attributes and collect more attributes as attribute repositories are queried.
+     * @param principal the "in progress" principal
+     */
+    public static void bindInProgressPrincipal(final Principal principal) {
+        val inProgressPrincipal = IN_PROGRESS_PRINCIPAL.get();
+        if (inProgressPrincipal != null) {
+            inProgressPrincipal.getAttributes().putAll(principal.getAttributes());
+        } else {
+            IN_PROGRESS_PRINCIPAL.set(principal);
+        }
```

Comment:
This does not seem safe; this is not just binding a principal but also modifying it. If we just want to expose a principal that is already resolved, sure, as that would be basically get and set ops. But this changes things that make it very difficult to troubleshoot or even maintain, specially the manual copying of the attributes. For example:

1. If the Principal interface is ever modified to include additional fields/flags, we'd have to track this area to make sure we are also copying more stuff...or realize why we don't or shouldn't have to.
2. It appears as though you are not really binding the principal, but you do so only to fetch attributes that are attached to it. Semantically, this feels strange. i.e., what is stopping an API call from binding a principal that is different from what's already tracked and bound, maybe even with a different identifier?

Reply:
Agreed that I shouldn't be using the Principal object, it was convenient b/c it had the attributes map I wanted, and I was undecided about whether to save the attributes map directly or introduce a new object/interface that only contained the attributes map. The attributes map gets passed as-is to attribute release policies where it is used in scripts so it sort of an API object on its own... 

I am trying to accumulate attributes including the ones from the authentication and make them available to subsequent attribute repositories and using a thread-local avoids having to add more APIs and complexity to person-directory where many of the attribute repositories would probably not benefit from the changes.